### PR TITLE
feat: upgrade pipeline

### PR DESCRIPTION
**Pull request by odyslam**
## Solution
- Forge script that deploys new implementations on multiple chains and updates the config file
- Forge script that reads the updated config file and produces the Upgrade callBatches, one for every chain. It can also produce the built combined callBatch in case we execute every upgrade on every chain independently (e.g recovery mode)
- The `actions` directory top-level directory is used for input/output files
- Requires: https://github.com/foundry-rs/foundry/pull/3647 and https://github.com/foundry-rs/forge-std/pull/213

## PR Checklist

- [ ] Added Tests
- [ ] Updated Documentation
- [ ] Updated CHANGELOG.md for the appropriate package

### CODE REVIEW

File: foundry.toml [hunk 1]
```diff
@@ -1,16 +1,16 @@
 [profile.default]
-solc_version = '0.7.6'        # Version of solc that we use
```
Comment (prestwich) [inline foundry.toml:2]: why all the formatting changes?

File: packages/contracts-ops/contracts/DeployImplementations.s.sol [hunk 2]
```diff
@@ -0,0 +1,221 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
```
Comment (prestwich) [inline packages/contracts-ops/contracts/DeployImplementations.s.sol:1]: file exists twice? once in `contracts/` and  once in `contracts/scripts/`

**Comment by prestwich** [comment]
as a procedural question, is there any way we can avoid deploying the implementation if it's a NOP? i.e. the implementation code would not change? it seems difficult as we use immutables 🤔

File: packages/contracts-ops/contracts/DeployImplementations.s.sol [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+    function title(string memory title1, string memory title2) internal {
+        console2.log(" ");
+        console2.log("===========================");
+        console2.log(title1, title2);
+        console2.log("===========================");
+        console2.log(" ");
+    }
+
+    function writeImpl() internal {
```
Comment (prestwich) [inline packages/contracts-ops/contracts/DeployImplementations.s.sol:156]: refactor to use JsonWriter?

File: packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        // initializeImplementations();
+        writeImpl();
+        vm.stopBroadcast();
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                      IMPLEMENTATION DEPLOYMENT
+  //////////////////////////////////////////////////////////////*/
+
+    function deployImplementations() internal {
```
Comment (prestwich) [inline packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol:91]: clean

File: packages/contracts-ops/contracts/scripts/RotateUpdater.s.sol [hunk 5]
```diff
@@ -82,6 +82,6 @@ contract RotateUpdater is Script, Config, CallBatch {
     ) public {
         initialize(configFile, localDomain, output);
         setUpdater();
-        build(address(governanceRouter(localDomain).proxy));
+        build(address(governanceRouterUpgrade(localDomain).proxy));
```
Comment (prestwich) [inline packages/contracts-ops/contracts/scripts/RotateUpdater.s.sol:85]: change should be rebased out

File: packages/contracts-ops/scripts/updateConfig.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
+import * as config from "@nomad-xyz/configuration";
```
Comment (prestwich) [inline packages/contracts-ops/scripts/updateConfig.ts:1]: I like this but think it needs more development. what are the common actions, and do we have standard formats for them?

I like having the implementations interface 🤔

File: packages/contracts-ops/contracts/scripts/UpgradeGoVBatches.s.sol [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+// Ops libs
+import {Config} from "../Config.sol";
+import {CallBatch} from "../Callbatch.sol";
+
+contract UpgradeGovBatches is Test, Config, CallBatch {
+    /*//////////////////////////////////////////////////////////////
+                                 BEACONS
+    //////////////////////////////////////////////////////////////*/
+
+    address homeBeacon;
```
[the remark this answers is not in the document]
Comment (odyslam) [reply packages/contracts-ops/contracts/scripts/UpgradeGoVBatches.s.sol:22]: What do you mean?

File: packages/contracts-ops/contracts/DeployImplementations.s.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (odyslam) [reply packages/contracts-ops/contracts/DeployImplementations.s.sol:156]: This is for console output, not file/json

File: foundry.toml (hunk 1 above)
Comment (odyslam) [reply foundry.toml:2]: vim

File: packages/contracts-ops/contracts/DeployImplementations.s.sol (hunk 2 above)
Comment (odyslam) [reply packages/contracts-ops/contracts/DeployImplementations.s.sol:1]: yy not sure why. I don't see this locally.

File: packages/contracts-ops/contracts/DeployImplementations.s.sol (hunk 3 above)
Comment (odyslam) [reply packages/contracts-ops/contracts/DeployImplementations.s.sol:156]: Yes

File: foundry.toml (hunk 1 above)
Comment (prestwich) [reply foundry.toml:2]: yeah, but can we not

File: packages/contracts-ops/contracts/CallBatch.sol [hunk 8]
```diff
@@ -44,92 +44,128 @@ abstract contract CallBatch is Script {
...
+
+    function writeObjectBody(string memory indent, string[2][] memory kvs)
+        private
+    {
+        for (uint256 i = 0; i < kvs.length; i++) {
+            writeKV(indent, kvs[i][0], kvs[i][1], i == kvs.length - 1);
+        }
+    }
+
+    function writeArrayOpen(string memory indent, string memory key) private {
```
Comment (anna-carroll) [inline packages/contracts-ops/contracts/CallBatch.sol:78]: why is this in CallBatch? needs to be rebased out, this should be in JSONWriter, no?

File: packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+            loadConfig();
+            vm.startBroadcast();
+            deployImplementations();
+            updateImpl(localDomain);
+            vm.stopBroadcast();
+        }
+    }
+
+    function loadConfig() internal {
+        recoveryTimelock = getGovernanceConfiguration(localDomain)
```
Comment (anna-carroll) [inline packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol:83]: just use calls to config inline, it's less boilerplate

File: packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+import {JsonWriter} from "../JsonWriter.sol";
+
+contract DeployImplementations is Test, Config {
+    using JsonWriter for JsonWriter.Buffer;
+
+    /*//////////////////////////////////////////////////////////////
+                            UPGRADE ARGUMENTS
+  //////////////////////////////////////////////////////////////*/
+
+    // configuration
```
Comment (anna-carroll) [inline packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol:31]: just use this from Config - no?

**Comment by odyslam** [comment]
We should first merge #516 , rebase this and then merge.  Review only the deployImplementations script

File: packages/contracts-ops/contracts/CallBatch.sol (hunk 8 above)
Comment (odyslam) [reply packages/contracts-ops/contracts/CallBatch.sol:78]: Yeah, ignore this file. It will be rebased.

**Comment by odyslam** [comment]
I had to break the fork creation and fork selection into two different phases because of a forge bug (that was just fixed in https://github.com/foundry-rs/foundry/pull/3679)

The script works as it should.

Next step is to open a PR with the upgradeCallbatches and update the readme on how all this should be used.

File: packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        // GovernanceRouter
+        governanceRouter = new GovernanceRouter(
+            localDomainNumber,
+            recoveryTimelock
+        );
+        governanceRouter.initialize(xAppConnectionManager, address(0));
+        console2.log("Governance Router deploed at", address(governanceRouter));
+
+        // BridgeRouter
+        bridgeRouter = new BridgeRouter();
```
Comment (anna-carroll) [inline packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol:152]: what about `EthereumBridgeRouter` for ethereum?

File: packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+        vm.writeJson(vm.toString(address(home)), configPath, valueKey);
+        valueKey = string(
+            abi.encodePacked(
+                ".core.",
+                domain,
+                ".governanceRouter",
+                ".implementation"
+            )
+        );
+        vm.writeJson(
```
Comment (anna-carroll) [inline packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol:180]: this does not compile

Comment (odyslam) [reply packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol:180]: It just needs the latest forge-std

Comment (odyslam) [reply packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol:180]: Which I made a different PR for

File: packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol (hunk 11 above)
Comment (anna-carroll) [reply packages/contracts-ops/contracts/scripts/DeployImplementations.s.sol:152]: @odyslam if we wanted to merge this at some point, would have to take this into account. but by now, i think we should move forward with my modifications of your work, as it's closer to mergeable. you can refactor / edit that work (And the other scripts) to make them multi-chain and use `writeJson` - but we shouldn't block timelines on that.

**Pull request closed without merging**